gen/decoder.go

```diff
@@ -170,7 +170,45 @@ func (g *Generator) genStructFieldDecoder(t reflect.Type, f reflect.StructField)
 	}
 
 	fmt.Fprintf(g.out, "    case %q:\n", jsonName)
-	return g.genTypeDecoder(f.Type, "out."+f.Name, tags, 3)
+	if err := g.genTypeDecoder(f.Type, "out."+f.Name, tags, 3); err != nil {
+		return err
+	}
+
+	if tags.required {
+		fmt.Fprintf(g.out, "%sSet = true\n", jsonName)
+	}
+
+	return nil
+}
+
+func (g *Generator) getFieldSetBlock(t reflect.Type, f reflect.StructField) error {
+	jsonName := g.namer.GetJSONFieldName(t, f)
+	tags := parseFieldTags(f)
+
+	if !tags.required {
+		return nil
+	}
+
+	fmt.Fprintf(g.out, "var %sSet bool\n", jsonName)
+
+	return nil
```

Comment:
Same as with getFieldCheckBlock: I'd not bother having an error at all if the method never fails.